common/check_expired_locked_rules

```diff
@@ -1,56 +1,123 @@
 #
-# Authors:
-# - Cedric Serfon, <cedric.serfon@cern.ch>, 2015
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
 
-'''
+"""
 Probe to check the locked expired rules or datasets with locked rules
-'''
+"""
 
 import sys
-from rucio.db.sqla.session import get_session
+import traceback
+from collections import defaultdict
+from datetime import datetime, timezone
+
+from rucio.db.sqla import models, session
+from sqlalchemy import and_, null, select, true
+from utils.common import PrometheusPusher
 
 # Exit statuses
 OK, WARNING, CRITICAL, UNKNOWN = 0, 1, 2, 3
 
-
-def main():
-    '''
-    Probe to check the locked expired rules or datasets with locked rules
-    '''
+if __name__ == "__main__":
     status = OK
-    session = get_session()
-    try:
-        query = "select rawtohex(id), scope, name, rse_expression from atlas_rucio.rules where locked=1 and expires_at<sys_extract_utc(localtimestamp)"
-        print 'Locked expired rules'
-        for row in session.execute(query):
-            status = CRITICAL
-            print row[0], row[1], row[2]
-    except Exception as error:
-        print error
-        status = UNKNOWN
-        sys.exit(status)
-    try:
-        query = """select rawtohex(c.id), c.scope, c.name, c.rse_expression from atlas_rucio.rules c,
-                   (select a.scope, a.name from atlas_rucio.dids a
-                   where a.expired_at is not null and a.expired_at < sys_extract_utc(localtimestamp)
-                   and exists (select 1 from atlas_rucio.rules b where a.scope=b.scope and a.name=b.name and locked=1)) d
-                   where c.scope=d.scope and c.name=d.name and locked=1"""
-        print 'Datasets expired with locked rules'
-        for row in session.execute(query):
-            status = CRITICAL
-            print row[0], row[1], row[2], row[3]
-    except Exception as error:
-        print error
-        status = UNKNOWN
-        sys.exit(status)
-    sys.exit(status)
+    session = session.get_session()
 
+    # Select statement used for both metrics
+    base_statement = select(
+        models.ReplicationRule,
+        models.DataIdentifier.name,
+        models.DataIdentifier.scope,
+        models.ReplicationRule.rse_expression,
+    )
+    # Use prometheus pusher to send results to a remote service
+    with PrometheusPusher() as manager:
+        try:
+            statement = base_statement.where(
+                and_(
+                    models.ReplicationRule.locked == true(),
+                    models.ReplicationRule.expires_at < datetime.now(tz=timezone.utc),
+                )
+            )
+            query = session.execute(statement).all()
```

Comment:
Sounds like a plan. Should we update the sql2.0 guidelines in the docs to reflect this?

Reply:
In due time. It is envisioned to have an entire subsection and I’ve taken some notes over the past months, but I have to review them carefully so that we can trust that the guidelines are sound.

> The Rucio project has adopted a particular coding style for its interaction with the database. It can be split into two parts: constructing the SQL statement, and executing it and handling its results.